xmppserver/src/main/java/org/jivesoftware/util/cache/ExternalizableUtil.java

```diff
@@ -35,14 +41,37 @@
  */
 public class ExternalizableUtil {
 
-    private static ExternalizableUtil instance = new ExternalizableUtil();
+    private static ExternalizableUtil instance;
 
     private ExternalizableUtilStrategy strategy = new DefaultExternalizableUtil();
 
     static {
         instance = new ExternalizableUtil();
     }
 
+    /**
+     * Pool of SAX Readers. SAXReader is not thread safe so we need to have a pool of readers.
+     */
+    private final static int POOL_SIZE = 10;
+    private final BlockingQueue<SAXReader> xmlReaders = new LinkedBlockingQueue<>(POOL_SIZE);
+
+    private SAXReader takeSAXReader() throws InterruptedException {
+        SAXReader xmlReader;
+        if (! xmlReaders.isEmpty()) {
+            xmlReader = new SAXReader();
+            xmlReader.setEncoding("UTF-8");
```

Comment:
This needs to have a couple of features set, to prevent some vulnerabilities to be exposed.

```
xmlReader.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
xmlReader.setFeature("http://xml.org/sax/features/external-general-entities", false);
xmlReader.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
```